src/connector/codec/src/decoder/json/mod.rs

```diff
@@ -12,11 +12,170 @@
+use url::Url;
 
 use super::avro::{avro_schema_to_column_descs, MapHandling};
 
+#[derive(Debug, Error, thiserror_ext::ContextInto)]
+pub enum Error {
+    #[error("could not open schema from {filename}")]
+    SchemaFromFile {
+        filename: String,
+        source: std::io::Error,
+    },
+    #[error("parse error for url {url}")]
+    UrlParse {
+        url: String,
+        source: url::ParseError,
+    },
+    #[error("schema from {url} not valid JSON")]
+    SchemaNotJson { url: String, source: std::io::Error },
+    #[error("request error")]
+    Request { url: String, source: reqwest::Error },
+    #[error("schema from {url} not valid JSON")]
+    SchemaNotJsonSerde {
+        url: String,
+        source: serde_json::Error,
+    },
+    #[error("ref `{ref_string}` can not be resolved as pointer, `{ref_fragment}` can not be found in the schema")]
+    JsonRefPointerNotFound {
+        ref_string: String,
+        ref_fragment: String,
+    },
+    #[error("json ref error")]
+    JSONRef {
+        #[from]
+        source: std::io::Error,
+    },
+    #[error("need url to be a file or a http based, got {url}")]
+    UnsupportedUrl { url: String },
+    #[error(transparent)]
+    Uncategorized(
+        #[from]
+        #[backtrace]
+        anyhow::Error,
+    ),
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Debug)]
+pub struct JsonRef {
+    schema_cache: HashMap<String, Value>,
+    reference_key: Option<String>,
+}
+
+impl JsonRef {
+    fn new() -> JsonRef {
+        JsonRef {
+            schema_cache: HashMap::new(),
+            reference_key: Some("__reference__".to_owned()),
+        }
+    }
+
+    async fn deref_value(&mut self, value: &mut Value, id: String) -> Result<()> {
+        let id_url = Url::parse(&id).into_url_parse(&id)?.to_string();
+        self.schema_cache.insert(id_url.clone(), value.clone());
+        self.deref(value, id_url, &vec![]).await?;
+        Ok(())
+    }
+
+    async fn deref(
+        &mut self,
+        value: &mut Value,
+        id: String,
+        used_refs: &Vec<String>,
+    ) -> Result<()> {
+        if let Some(obj) = value.as_object_mut() {
+            if let Some(ref_value) = obj.remove("$ref") {
+                if let Some(ref_string) = ref_value.as_str() {
+                    let id_url = Url::parse(&id).into_url_parse(&id)?;
+                    let ref_url = id_url.join(ref_string).into_url_parse(ref_string)?;
+                    let mut ref_url_no_fragment = ref_url.clone();
+                    ref_url_no_fragment.set_fragment(None);
+                    let url_schema = ref_url_no_fragment.scheme();
+                    let ref_no_fragment = ref_url_no_fragment.to_string();
+
+                    let mut schema = match self.schema_cache.get(&ref_no_fragment) {
+                        Some(cached_schema) => cached_schema.clone(),
+                        None => {
+                            if url_schema == "http" || url_schema == "https" {
```

Comment:
Are you sure we want to support schema reference from http/fs?
Should we support retrieve references from schema registry instead? https://docs.confluent.io/platform/current/schema-registry/fundamentals/serdes-develop/serdes-json.html#schema-references-in-json-schemas

Reply:
Is this feature a customer request? Please provide more context, thanks.

Reply:
User only requested schema ref resolution without an external link.

I thought the HTTP support can be a general approve to support schema registry. But I just realized there is a version segment after the subject. I'll change this pr to only support schema resolution without external resource for now.

Reply:
Actually, we do allow users using http and file. So it won't hurt to add those support.

It's just for the schema registry, we need to investigate more into how it works.

Reply:
> they provide a way to load schemas into an internal schema database

It's true a for a deref library to focus on derefing instead of fetching (in fact it's  `[referencing](https://github.com/Stranger6667/jsonschema/tree/master/crates/jsonschema-referencing)` do). But if we want do do so we will have to ask users to provide all the URL in the schema ref or parse the schema to get all the $ref recurrently before ref.